### DISCUSSION TITLE
Invoke Method

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#6115] [General] [answered]
I'd like to build an Event<TDelegate> class that wraps a list of TDelegates and exposes a custom Invoke method that:
1.  matches the signature of TDelegate
2.  Does some pre/post processing on the list of delegates. 

How do I accomplish #1?

**Comment by jnm2** [comment, marked as answer]
I've done (1) by exposing `public TDelegate Invoke { get; }`. The language doesn't provide a facility to change a method signature based on the signature of a delegate type parameter.

**Comment by HaloFour** [comment]
Is there an ask for a language feature here?

**Comment by TahirAhmadov** [comment]
Your three options are either 1) create a whole bunch of versions for each delegate signature (which is lots of repetition) or 2) use reflection to invoke the `Invoke` (pun intended) - basically takes `params object[] args` and passes it on (this is not type-safe) or 3) if possible, pass a delegate invocation delegate in: `events.Run(d=>d(x, y, z))`
